### PR TITLE
Add ProGuard rules fixes #11891 #6624

### LINKED ISSUE #11891: 0.41.0-rc.0 - Update Android Proguard Rules
### Description

Recently I updated to 0.41.0-rc.0 and there has been a regression in regards to Proguard for Android. The following happens on a release build:

```
Warning: com.facebook.fbui.textlayoutbuilder.proxy.StaticLayoutProxy:
can't find referenced method 'StaticLayout(java.lang.CharSequence,int,int,android.text.TextPaint,int,android.text.Layout$Alignment,android.text.TextDirectionHeuristic,float,float,boolean,android.text.TextUtils$TruncateAt,int,int)'
in library class android.text.StaticLayout
```

### Reproduction

1. Create a new project: `react-native init TestProguardProject`
2. Setup release as documented and run it: Must use --configuration flag rather then variant for some reason based on #11443 
3. Upgrade it to v41: `react-native-git-upgrade 0.41.0-rc.0`
4. Try running the release build again: `react-native run-android --variant=release`
5. Experience a build failure based on the above

### Solution

* It seems like a simple `-dontwarn` addition to proguard or better yet a `-keep` on the needed method would be best. Not a proguard expert so I'll leave this up for discussion, but can submit a PR if needed. 
* For reference: https://www.guardsquare.com/en/proguard/manual/troubleshooting#unresolvedlibraryclassmember

### Additional Information

* React Native version: 0.41.0-rc.0
* Platform: Android
* Operating System: MacOS

### PR DESCRIPTION
**Pull request by iegik**
Trying to build project for Android 7 (O)

## Test Plan

RN 0.47.1

```
        classpath 'com.android.tools.build:gradle:2.3.3' // For Android 7 (O)
```

```
    compile "com.android.support:appcompat-v7:26.0.0-alpha1"
```

```
    buildTypes {
        debug {
            debuggable true
            minifyEnabled enableProguardInReleaseBuilds
            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro", "devsupport.pro"
            applicationIdSuffix ".debug"
        }
        release {
            signingConfig signingConfigs.release
            minifyEnabled enableProguardInReleaseBuilds
            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
            manifestPlaceholders = [excludeSystemAlertWindowPermission: "true"]
        }
    }

```

```
    supportLibVersion = '26.0.1'
...
    compileSdkVersion 26
```

### CODE REVIEW

**Comment by akshetpandey** [comment]
This doesn't fix the crash for me with proguard enabled.

**Comment by akshetpandey** [comment]
Its because it should be this, the package names changed...
```patch
--keep,includedescriptorclasses class com.facebook.react.cxxbridge.ModuleRegistryHolder { *; }
--keep,includedescriptorclasses class com.facebook.react.cxxbridge.CatalystInstanceImpl { *; }
--keep,includedescriptorclasses class com.facebook.react.cxxbridge.JavaScriptExecutor { *; }
+-keep,includedescriptorclasses class com.facebook.react.bridge.CatalystInstanceImpl { *; }
+-keep,includedescriptorclasses class com.facebook.react.bridge.JavaScriptExecutor { *; }
```

**Comment by iegik** [comment]
@akshetpandey , can You check this?

File: local-cli/templates/HelloWorld/android/app/proguard-rules.pro [hunk 1]
```diff
@@ -68,3 +68,23 @@
...
+
+# RN 0.41+
+# https://github.com/facebook/react-native/issues/11891
+# -dontwarn com.facebook.fbui.textlayoutbuilder.proxy.StaticLayoutProxy
+# This library uses a non-public Android constructor within StaticLayout.
+# See libs/proxy/src/main/java/com/facebook/fbui/textlayoutbuilder/proxy for details.
+-dontwarn android.text.StaticLayout
+
+# https://github.com/facebook/react-native/issues/6624
+-keep class com.facebook.react.bridge.ModuleRegistryHolder { *; }
```
Comment (akshetpandey) [inline local-cli/templates/HelloWorld/android/app/proguard-rules.pro]: ModuleRegistryHolder isn't a class anymore.
Is it here for backwards compatibility?

**Comment by akshetpandey** [comment]
This works for me, I did leave a comment about an extra line which doesn't seem to do anything though.

**Comment by hramos** [comment]
Looks like some new changes were added to the PR? This invalidates the prior approval.

**Comment by akshetpandey** [comment]
At this point this should probably be changed to target 0.49/0.50 instead

**Comment by hramos** [comment]
Ah! That explains why this is not landing. We can only merge pull requests opened against master.

To land commits into stable releases, a release coordinator has to cherry pick the individual commit. Commenting on #16257 like you just did is the recommended approach. I'm going to close this as we can't land this PR.

**Pull request closed without merging**